### DISCUSSION TITLE
Something is eating my messages!

### DISCUSSION THREAD

**Discussion by noahzn** [bitcraze/discussions#822] [Q&A] [answered]
Hi, I am sending messages from GAP8 to the APP layer, and I also print cpx messages to the console. But I found that sometimes the first 6 letters of the message "veloci" had been eaten! Also, in the app layer I use DEBUG_PRINT to print messages, and the "recei" of the word "received" had been eaten as well.
![image](https://github.com/bitcraze/discussions/assets/8244850/267990b1-d9c8-4848-9e92-f1c2048a0cfc)

```
while (1) {   
    cpxPrintToConsole(LOG_TO_CRTP, txp2.dataLength);
    createVelMsgPacket(&txp2);
    cpxSendPacketBlocking(&txp2);
    vTaskDelay(2000);
    cpxPrintToConsole(LOG_TO_CRTP, "velocity has been sent!\n");
    pi_time_wait_us(1000*1000);
  }
```
Do you know what happened? Thanks!

**Comment by ataffanel** [comment]
This is very odd, there seems to be a protocol mismatch somewhere. Everyone with experience with the AI-deck here at Bitcraze is currently on Vacation but I can try to help a bit.

Does it happen all the time or only some time? Is there any logic or anything happening between the time it work and the time it does not?

**Comment by noahzn** [comment]
Hi, yes, this happens every time. My code is based on the face detection demo, and I added a parallel task (send_msg) to send messages to the app layer. The problematic cpx printing function is in the app layer and in the `send_msg` task.

**Comment by ataffanel** [comment]
I am not exactly sure what is happening but one thing I can see is that the `cpxPrintToConsole` function is not thread safe, it is using global context so it should not be called from multiple thread at the same time.

Is it possible for you to try a simpler hello-world on your deck to find out if this bug is related to your program or is it a system bug that is also present in simple hello-worlds?

**Reply by noahzn** [reply]
Thank you for your reply. I tried only printing simple messages in the hello-world app. And it can print without issues, but at the same time, the "sending" function on the GAP8 doesn't work properly. It doesn't print the message "velocity has been sent!\n". I think this is caused by using `cpxSendPacketBlocking`. If the app layer doesn't receive the message, then it's blocked, do I understand correctly?
```
while (1) {   
    cpxPrintToConsole(LOG_TO_CRTP, txp2.dataLength);
    createVelMsgPacket(&txp2);
    cpxSendPacketBlocking(&txp2);
    vTaskDelay(2000);
    cpxPrintToConsole(LOG_TO_CRTP, "velocity has been sent!\n");
    pi_time_wait_us(1000*1000);
  }
```

**Reply by krichardsson** [reply, marked as answer]
yes, CPX is generally like a pipe, if you push packets in the GAP8 and they are not popped in the other end (the STM in this case) the GAP8 will be blocked when all queues int path have filled up. However the intention is that packets should be popped and this should not happen. 

In the code above `cpxPrintToConsole()` is also calling `cpxSendPacketBlocking()` so in essence you have 3 calls to `cpxSendPacketBlocking()`, but the packets will be handled by different code in the STM. 

The logging packets are handled by the [`CPX_F_CONSOLE` case](https://github.com/bitcraze/crazyflie-firmware/blob/e56b7fddf6e6ac9e10d9bacfffb4c4bfcd629219/src/modules/src/cpx/cpx.c#L121-L129) in the function switcher. This code calls `DEBUG_PRINT()` which should be fast (and not a problem).

The other packet that you send in `cpxSendPacketBlocking(&txp2);` I do not now how it is defined. If you are using the `CPX_F_APP` it will be handled by the [`CPX_F_APP` case](https://github.com/bitcraze/crazyflie-firmware/blob/e56b7fddf6e6ac9e10d9bacfffb4c4bfcd629219/src/modules/src/cpx/cpx.c#L154-L157), and if you have registered an app message handler it will be called. If this handler takes a long time to execute you will be blocked in the GAP8.
If no handler is registered, the packet will simply be dropped. 


The above discussion does not really explain why part of your logging is missing though.

I also noticed that there is actually more than "veloci" that has been eaten, there should be some more info added by the logging framework ("CPX: ") and the CPX handler ("GAP8: ") so I think "CPX: GAP8: veloci" is the full missing part. To me this indicates that the problem is related to the logging in the STM rather than a CPX problem. 

Another thing I notice is that the first log after the log in the app (the receive log) is OK, while the other logs are bad. This could indicate that a log message is truncated if it is printed **after** a log message from the GAP8, and that the real problem is the previous message. Maybe there is a problem with the ending for some reason? An interesting experiment could be to add a "\n" when the GAP8 message is logged in the STM, just to see what happens.
Change [`DEBUG_PRINT("GAP8: %s", cpxRx.data`);](https://github.com/bitcraze/crazyflie-firmware/blob/e56b7fddf6e6ac9e10d9bacfffb4c4bfcd629219/src/modules/src/cpx/cpx.c#L125) to `DEBUG_PRINT("GAP8: %s\n", cpxRx.data`

**Reply by noahzn** [reply]
Dear @krichardsson, thank you very much for your detailed explanation. Yes, I am using an app message handler. Currently, the handler only prints the fake data sent by GAP8. 

 I changed the code in `cpx.c` as you mentioned, and now it works perfectly. 

![image](https://github.com/bitcraze/discussions/assets/8244850/da8f67bc-888f-4c3f-85a2-e91836dddaea)

Thank you again for your help! Also, thanks to @ataffanel

**Comment by krichardsson** [comment]
Great that it solved your problem! 

There are some details in this though that looks a bit suspicious to me and I have created an [issue](https://github.com/bitcraze/crazyflie-firmware/issues/1304) for further investigations.
